### DISCUSSION TITLE
when trying to update records of my table it is giving me these errors

### DISCUSSION THREAD

**Discussion by anish6701** [fastapi/fastapi#8507] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
This is the user class which is making my routes, I am getting errors in my Put method. My function update has issues and I was following this on youtube by bitfumes.

from fastapi import APIRouter,Depends,status,HTTPException
from  MerchantRegistration import models,database,hashing,schemas,oauth2
from sqlalchemy.orm import Session
from typing import List
from repository import user

router=APIRouter(
    prefix="/user",
    tags=['Users']
)

@router.get('/', response_model=List[schemas.ShowUsr])
def all(db:Session=Depends(database.get_db),current_user:schemas.UsrBase=Depends(oauth2.get_current_user)):
    users=db.query(models.User).all()
    return users

@router.post('/',status_code=status.HTTP_201_CREATED,response_model=schemas.ShowUsr)
def create(request:schemas.UsrBase, db:Session=Depends(database.get_db)):
    Lst_of_users= models.User(user_name=request.user_name, f_name=request.f_name,
                             l_name=request.l_name, email=request.email, phonenumber=request.phonenumber,
                             companyName=request.companyName, websiteURL=request.websiteURL, password=hashing.Hash.bcrypt(request.password)
        )

    db.add(Lst_of_users)
    db.commit()
    db.refresh(Lst_of_users)
    return Lst_of_users

@router.put('/{id}',status_code=status.HTTP_202_ACCEPTED,response_model=schemas.ShowUsr)
def update(id:int,request:schemas.UsrBase,db:Session=Depends(database.get_db)):

    print(id)
    usr = db.query(models.User).filter(models.User.user_id == id)
    print(usr)
    if not usr.first():
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND,
                            detail  = f"User with id {id} is not available ")
    usr.update(request)
    db.commit()
    return 'done'
  

@router.delete('/{id}',status_code=status.HTTP_204_NO_CONTENT)
def destroy(id,db:Session=Depends(database.get_db),current_user:schemas.UsrBase=Depends(oauth2.get_current_user)):
    usr=db.query(models.User).filter(models.User.user_id == id)
    if not usr.first():
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND,
                            detail=f"User with id {id} is not available ")

    usr.delete(synchronize_session=False)
    db.commit()
    return 'done'

@router.get('/{id}',response_model=schemas.ShowUsr,status_code=200)
def get_user(id,db:Session=Depends(database.get_db),current_user:schemas.UsrBase=Depends(oauth2.get_current_user)):

    user=db.query(models.User).filter(models.User.user_id == id).first()
    print(user)
    if not user:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND,
                            detail = f"User with id {id} is not available ")
        # response.status_code=status.HTTP_404_NOT_FOUND
        # return {"detail" : f"User with id {id} is not available "}

    return user


My model is below 

from sqlalchemy import Column,Integer,String
from database import Base

class User(Base):
    __tablename__='user'
    __table_args__ = {'extend_existing': True}

    user_id=Column(Integer,primary_key=True, index=True)
    user_name=Column(String)
    f_name=Column(String)
    l_name=Column(String)
    email=Column(String)
    phonenumber=Column(String)
    websiteURL=Column(String)
    websiteDetails=Column(String)
    companyName=Column(String)
    password=Column(String)
    howYouHearAboutUs=Column(String)


The Schemas is defined below : 

from pydantic import BaseModel,Field
from typing import List,Optional

class UsrBase(BaseModel):
    user_name:str
    f_name:str
    l_name: str
    email:str
    phonenumber: str
    websiteURL:str
    websiteDetails: str
    companyName: str
    password:str
    howYouHearAboutUs: str

class Usr(UsrBase):
    class Config():
        orm_mode=True

class ShowUsr(BaseModel):
    user_id:str
    user_name:str
    companyName:str
    email:str
    websiteURL:str
    password:str

    class Config():
        orm_mode=True



class Login(BaseModel):
    username:str
    password:str

class Token(BaseModel):
    access_token:str
    token_type:str

class TokenData(BaseModel):
    email: Optional[str] = None
```

### Description

when I try to update a record in my DB this is the error which I am getting : 

SELECT "user".user_id AS user_user_id, "user".user_name AS user_user_name, "user".f_name A
S user_f_name, "user".l_name AS user_l_name, "user".email AS user_email, "user".phonenumbe
r AS user_phonenumber, "user"."websiteURL" AS "user_websiteURL", "user"."websiteDetails" A
S "user_websiteDetails", "user"."companyName" AS "user_companyName", "user".password AS user_password, "user"."howYouHearAboutUs" AS "user_howYouHearAboutUs"
FROM "user"
WHERE "user".user_id = **%(user_id_1)s**

I am printing my query in the put method so that I can check whether the parameters are being matched or not but in the where clause why is the "%" coming before the user_id and why _1 is appended to it ? 

### Operating System

Windows

### FastAPI Version

0.78.0

### Python Version

3.10.2

**Comment by slackner** [comment]
The `%(user_id_1)s` part of your SQL query looks correct to me. This is a just a placeholder, when executing the query SQLAlchemy will insert the corresponding values or pass them separately to the database.

When looking at the code, first thing I noticed is that `id` does not have any type in some API endpoints. This means you will get the id as a string (e.g. in `destroy` and `get_user`), and your database will probably not know how to deal with the string - integer comparison.

In addition, the `update` method passes a pydantic object to SQLAlchemy in `usr.update(request)`. I suspect it should be something like `usr.update(request.dict())`, since the underlying [update](https://docs.sqlalchemy.org/en/14/orm/query.html#sqlalchemy.orm.Query.update) method expects a dictionary.

Do these hints solve your issue? If not, please provide a full Python backtrace of the exception you are encountering.

**Comment by anish6701** [comment]
Hi as you mentioned above I have added the data type for **id** and also updated the **usr.update(request.dict())**. After making these changes the data when being altered on the swagger UI, **It is getting updated in the database** but still gives an **Internal server error**.
Below I have attached the screenshots of the browser and the error which I got on my terminal.

@slackner  would be great if you could help me out with this error.

For my reference I learned FastAPI from this youtube link 
https://www.youtube.com/watch?v=7t2alSnE2-I&t=11051s

and started to make my own project using his learning. 






![Screenshot 2022-05-31 114228](https://user-images.githubusercontent.com/60069224/171104929-e59467cc-ac4d-45e0-9ebe-c0464938c350.png)

![Screenshot 2022-05-31 114006](https://user-images.githubusercontent.com/60069224/171104653-1421f93a-db3b-411b-beaf-97d339c64b62.png)

**Comment by slackner** [comment]
The traceback still seems to be truncated, it lacks information about where it crashes on your Python code. Can you share the full traceback please? If its too long to include here, you can also use a [gist](https://gist.github.com/) and just share the link.

**Comment by anish6701** [comment]
Hi, @slackner I updated my put method in this way (posting the code below) and all my fields are being inserted into the database too. 

The repository has been defined here: 

# Updating the user details that they would like to update
def update(id:int,request:schemas.UsrBase,db:Session):
    p = request.password
    request.password = hashing.Hash.bcrypt(p)
    user = db.query(models.User).filter(models.User.user_id == id)
    if not user.first():
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND,
                            detail=f'User with id {id} is not available')
    user.update(request.dict())
    db.commit()
    return 'updated'

The code above which I had raised my ticket has a response_model, which I removed in my new code. It worked properly and did not throw any errors.


The router is defined here:

@router.put('/{id}',status_code=status.HTTP_202_ACCEPTED)
def update(id:int,request:schemas.UsrBase,db:Session=Depends(database.get_db),current_user:schemas.UsrBase=Depends(oauth2.get_current_user)):
    return user.update(id,request,db)

**Comment by anish6701** [comment]
Thank you @slackner for your input which helped me to solve my error. 

I had a new doubt about which I have declared my column as a JSON data type and passed a dict to it which is storing certain fields of data to it. The screenshots below would help you to understand my concern 

# My routes for the program

![str1](https://user-images.githubusercontent.com/60069224/171604754-1b989a03-3bb6-4bfe-9b10-8350817dc0c2.png)
 
# My schemas for the program 

![str2](https://user-images.githubusercontent.com/60069224/171604906-6e17f083-3e66-41b0-82e5-ab795d5126c3.png)

# My model for the program 

![str3](https://user-images.githubusercontent.com/60069224/171604993-0ff74ff8-aa71-4d97-8334-8c0166c07670.png)

# My swagger UI where I am inputting the data 

![str4](https://user-images.githubusercontent.com/60069224/171605122-17c9473e-d600-4ed6-bc3e-6830d7d93b50.png)

So in the swagger UI as you can see the fields like **name, s_name, work_day, work_email** I want those variables to be constant and the values would be changing all time whenever the post or put method will be called.
